### PR TITLE
[#224] Implements B3 single header

### LINKED ISSUE #224: implement "b3 single" header format
## Issue

We should permit at least parsing the more compressed "b3" header format, if not making it possible to propagate it downstream.

ex.  `b3: 4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-1`

This is the same format as will be used in the w3c tracestate

## Context

Details are here: https://github.com/openzipkin/b3-propagation/issues/21

Example implementation is here: https://github.com/openzipkin/brave/blob/master/brave/src/main/java/brave/propagation/B3SingleFormat.java https://github.com/openzipkin/brave/blob/master/brave/src/test/java/brave/propagation/B3SingleFormatTest.java

### PR DESCRIPTION
**Pull request by jcchavezs**
This PR adds support for [b3 single header](https://github.com/openzipkin/b3-propagation#single-header). Closes #224 

* [x] I have viewed signed and have submitted the Contributor License Agreement

* [x] I have made this pull request to the `master` branch

* [x] I have run all the unit tests using `bin/test`

* [ ] I have run CF Acceptance Tests on bosh lite

### CODE REVIEW

**Comment by jcchavezs** [comment]
@KauzClay there is no way for me to run the tests. I receive the message:

`
../../../../src/code.cloudfoundry.org/gorouter/logger/lager_adapter.go:7:2: code in directory /Users/josecarloschavez/Workspace/gource/src/github.com/uber-go/zap expects import "go.uber.org/zap"
`

Any idea on how to sort out this problem?

**Comment by KauzClay** [comment]
Hi @jcchavezs , it's a little hard to tell from just that line, but it seems like an import issue. My initial suggestion would be to take a look at [this](https://golang.org/cmd/go/#hdr-Import_path_checking) and see if you see any conflicts with that in this file `/Users/josecarloschavez/Workspace/gource/src/github.com/uber-go/zap`
Let us know if you find anything!

**Comment by jcchavezs** [comment]
@KauzClay could you please enable travis for foreign branches?

**Comment by jcchavezs** [comment]
@KauzClay I actually made it work by using `zap` in version `v0.1.0-beta.1`. I will sign the CLA and then we just need this branch to have the test run. cc @adriancole

File: common/http/headers_test.go [hunk 1]
```diff
@@ -11,9 +11,6 @@ import (
 
 const uuid_regex = `^[[:xdigit:]]{8}(-[[:xdigit:]]{4}){3}-[[:xdigit:]]{12}$`
 
-// 64-bit random hexadecimal string
-const b3_id_regex = `^[[:xdigit:]]{16}$`
```
Comment (jcchavezs) [inline common/http/headers_test.go]: not used anywhere (and it is private)

**Comment by adriancole** [comment]
if you want to do an integration test, you can do against https://github.com/openzipkin/brave-webmvc-example which has most recently patched b3 single goodness

**Comment by adriancole** [comment]
by integration I mean a manual one, not saying you hook the builds

**Comment by adobley** [comment]
@jcchavezs Have you signed the CLA? The commit looks good to me, but you will need to close and re-open the commit so that dreddbot can take another look and verify that everything is OK.

**Comment by jcchavezs** [comment]
@adobley I just reopened the PR, not sure if I have to push the branch to the main repo so the CI gets triggered.

**Comment by jcchavezs** [comment]
@adobley could you please check the status on this?

**Comment by jeffpak** [comment]
Hello @jcchavezs ,

We're wondering if you've run CATs on this. We're taking a look at this PR again, but just wanted to double check and see if acceptance tests are passing. 

Sincerely,
CF Routing Team

**Comment by jcchavezs** [comment]
@jeffpak do you mean "I have run CF Acceptance Tests on bosh lite"? If that is the case I did not. I was not sure that was a thing (since [bosh lite repo says it is deprecated](https://github.com/cloudfoundry/bosh-lite)), is there any link or documentation I could look at?

**Comment by tstannard** [comment]
Hi @jcchavezs, 

I tried to run the gorouter tests on your PR and they failed :( 


```There were failures detected in the following suites:
        proxy ./proxy

Ginkgo ran 29 suites in 3m15.639061428s
Test Suite Failed
+ printStatus
+ '[' 1 -eq 0 ']'
+ echo -e '\nSUITE FAILURE'

SUITE FAILURE```

File: common/http/headers_test.go (hunk 1 above)
Comment (bruce-ricard) [reply common/http/headers_test.go]: Great. Please try to do that in a separate commit next time.

**Comment by bruce-ricard** [comment]
@jcchavezs we really appreciate the refactoring, thank you for that. Please next time consider making a first refactoring commit, and then the one with the actual functional changes. It makes it so much easier to code review.

**Comment by bruce-ricard** [comment]
We were able to rebase from cloudfoundry/master and run all tests, which passed.

@shubhaat @shalako we'd like to discuss with (one of) you if you're happy with the functional change this brings.

**Comment by shalako** [comment]
I've reviewed the value of the new header from docs at https://cwiki.apache.org/confluence/display/ZIPKIN/b3+single+header+format. 

@jcchavezs I assume the behavior you've contributed is that when tracing support is enabled in gorouter, the `b3` header will be forwarded and logged in access log messages? Will gorouter also generate the `b3` header based on known traceid, span, and parentspan?

**Comment by jcchavezs** [comment]
@shalako I changed the implementation so when receiving the b3 headers we still deliver the single version.

**Comment by jcchavezs** [comment]
@shalako @bruce-ricard any chance we can merge this?

**Comment by shubhaat** [comment]
@jcchavezs - sorry for the delay. I have pulled this up in the backlog, and hopefully we will get to merging this today. 

Thank you!

**Comment by adobley** [comment]
Going to merge and bump the submodule in routing-release.

Thanks a ton for the PR! 🎉

**Pull request merged**